PyAlbany/python/RandomizedCompression.py

```diff
@@ -0,0 +1,342 @@
+# ***********************************************************************
+# @HEADER
+
+# This file contains routines for randomized truncated singular value decompositions
+# of a matrix free operator (Op), by singlePass and doublePass algorithms.
+# A routine is available for the hierarchical off-diagonal low-rank (HODLR) decomposition
+# of a matrix free operator (Op). 
+#
+# See "Finding structure with randomness, probabilistic algorithms for constructing approximate
+# matrix decompositions", N. Halko, P.G. Martinsson, J.A. Tropp SIAM Review (2011),
+# "Compressing rank-structured matrices via randomized sampling", P.G. Martinsoon, SIAM Journal
+# on Scientific Computing (2016) and "Hierarchical off-diagonal low-rank approximation of
+# Hessians in inverse problems", T. Hartland, G. Stadler, M. Perego, K. Liegeois, N. Petra (to be submitted).  
+#
+# Various helper functions are utilized to allow for a parallel implementation of the HODLR compression
+# algorithm, such as Hpartition, MPIpartition and HidxMPIproc
+#
+#
+#
+# Last updated on April 19th, 2022
+# Author Tucker Hartland, University of California, Merced
+
+
+
+from PyTrilinos import Tpetra
+from PyTrilinos import Teuchos
+try:
+    from PyAlbany import wpyalbany as wpa
+except:
+    import wpyalbany as wpa
+
+try:
+    from PyAlbany import Utils as utils
+except:
+    import Utils as utils
+
+import numpy as np
+import scipy.linalg as spla
+
+
+"""
+  singlePass: randomized singular value decomposition of a symmetric operator
+  input: Op, operator which contains a Map member data and a 
+                  dot method which takes a Tpetra MultiVector
+                             and returns a Tpetra MultiVector 
+          k, the rank of the output operator U \Lambda U^T
+  output: \Lambda, U
+          \Lambda, eigenvalues,  a nondistributed numpy array with k entries
+          U,       eigenvectors, a distributed Tpetra MultiVector with k columns
+"""
+def singlePass(Op, k):
+    comm       = Teuchos.DefaultComm.getComm()
+    rank       = comm.getRank()
+    nprocs     = comm.getSize()
+    nElems     = Op.Map.getNodeNumElements()
+    N          = Op.Map.getGlobalNumElements()
+    nMaxOrthog = min(50, k)
```

Comment:
Can you precise in your comment about `k` that it cannot be above 50?

Reply:
I recall an issue when using `wpa.orthogTpMVecs`, when using more than 50 vectors at a time, that is orthogonalizing more than 50 vectors at once. Here it is reduced so that we attempt to orthogonalize the k columns using no more than 50 columns. I should look back into the orthogonalization at some point though.